File: packages/typescript/src/features/advanced-capabilities/clipboard/actions/list.ts
import {
	type Action,
	type HandlerCallback,
	type HandlerOptions,
	type IAgentRuntime,
	logger,
	type Memory,
	type State,
} from "../../../../types/index.ts";
import { createClipboardService } from "../services/clipboardService.ts";
import { requireActionSpec } from "../specs.ts";

const spec = requireActionSpec("CLIPBOARD_LIST");

export const clipboardListAction: Action = {
	name: spec.name,
	similes: spec.similes ? [...spec.similes] : [],
	description: spec.description,

	validate: async (
		runtime: IAgentRuntime,
		message: Memory,
		state?: State,
		options?: HandlerOptions,
	): Promise<boolean> => {
		const __avTextRaw =
			typeof message?.content?.text === "string" ? message.content.text : "";
		const __avText = __avTextRaw.toLowerCase();
		const __avKeywords = ["clipboard", "list"];
		const __avKeywordOk =
			__avKeywords.length > 0 &&
			__avKeywords.some((kw) => kw.length > 0 && __avText.includes(kw));
		const __avRegex = /\b(?:clipboard|list)\b/i;
		const __avRegexOk = __avRegex.test(__avText);
		const __avSource = String(message?.content?.source ?? "");
		const __avExpectedSource = "";
		const __avSourceOk = __avExpectedSource
			? __avSource === __avExpectedSource
			: Boolean(__avSource || state || runtime?.agentId || runtime?.getService);
		const __avOptions = options && typeof options === "object" ? options : {};
		const __avInputOk =
			__avText.trim().length > 0 ||
			Object.keys(__avOptions as Record<string, unknown>).length > 0 ||
			Boolean(message?.content && typeof message.content === "object");

		if (!(__avKeywordOk && __avRegexOk && __avSourceOk && __avInputOk)) {
			return false;
		}

		const __avLegacyValidate = async (
			_runtime: IAgentRuntime,
			_message: Memory,
		): Promise<boolean> => {
			return true;
		};
		try {
			return Boolean(await __avLegacyValidate(runtime, message));
		} catch {
			return false;
		}
	},

	handler: async (
		runtime: IAgentRuntime,
		message: Memory,
		_stateFromTrigger: State | undefined,
		_options: HandlerOptions | undefined,
		callback?: HandlerCallback,
		_responses?: Memory[],
	) => {
		try {
			const service = createClipboardService(runtime);
			const entries = await service.list();

			if (entries.length === 0) {
				if (callback) {
					await callback({
						text: "You don't have any clipboard entries yet. Use CLIPBOARD_WRITE to create one.",
						actions: ["CLIPBOARD_LIST_EMPTY"],
						source: message.content.source,
					});
				}
				return { success: true, text: "No entries", entries: [] };
			}

			const listText = entries
				.map((e, i) => {
					const tagsStr = e.tags?.length ? ` [${e.tags.join(", ")}]` : "";
					return `${i + 1}. **${e.title}** (${e.id})${tagsStr}\n   _Modified: ${e.modifiedAt.toLocaleDateString()}_`;
				})
				.join("\n");

			const successMessage = `**Your Clipboard Entries** (${entries.length} total):\n\n${listText}`;

			if (callback) {
				await callback({
					text: successMessage,
					actions: ["CLIPBOARD_LIST_SUCCESS"],
					source: message.content.source,
				});
			}

			return { success: true, text: successMessage, entries };
		} catch (error) {
			const errorMsg = error instanceof Error ? error.message : String(error);
			logger.error("[ClipboardList] Error:", errorMsg);
			if (callback) {
				await callback({
					text: `Failed to list clipboard entries: ${errorMsg}`,
					actions: ["CLIPBOARD_LIST_FAILED"],
					source: message.content.source,
				});
			}
			return { success: false, text: "Failed to list clipboard entries" };
		}
	},

	examples: [],
};

export default clipboardListAction;

File: packages/python/elizaos/advanced_capabilities/clipboard/actions/list.py
"""Clipboard list action -- list all clipboard entries."""

from __future__ import annotations

from typing import TYPE_CHECKING

from elizaos.types import Action, ActionResult, Content

from ..services.clipboard_service import create_clipboard_service

if TYPE_CHECKING:
    from elizaos.types import HandlerCallback, HandlerOptions, IAgentRuntime, Memory, State


async def _validate(runtime: IAgentRuntime, message: Memory, _state: State | None = None) -> bool:
    text = (message.content.text or "").lower()
    return "clipboard" in text and "list" in text


async def _handler(
    runtime: IAgentRuntime,
    message: Memory,
    state: State | None = None,
    options: HandlerOptions | None = None,
    callback: HandlerCallback | None = None,
    responses: list[Memory] | None = None,
) -> ActionResult:
    try:
        service = create_clipboard_service(runtime)
        entries = await service.list()

        if not entries:
            if callback:
                await callback(
                    Content(
                        text="You don't have any clipboard entries yet. Use CLIPBOARD_WRITE to create one.",
                    )
                )
            return ActionResult(text="No entries", success=True)

        lines = []
        for i, e in enumerate(entries):
            tags_str = f" [{', '.join(e.tags)}]" if e.tags else ""
            lines.append(
                f"{i + 1}. **{e.title}** ({e.id}){tags_str}\n"
                f"   _Modified: {e.modified_at.strftime('%Y-%m-%d')}_"
            )

        msg = f"**Your Clipboard Entries** ({len(entries)} total):\n\n" + "\n".join(lines)
        if callback:
            await callback(Content(text=msg, actions=["CLIPBOARD_LIST_SUCCESS"]))
        return ActionResult(text=msg, success=True)
    except Exception as e:
        if callback:
            await callback(Content(text=f"Failed to list clipboard entries: {e}"))
        return ActionResult(text="Failed to list clipboard entries", success=False)


clipboard_list_action = Action(
    name="CLIPBOARD_LIST",
    similes=["LIST_NOTES", "SHOW_NOTES", "CLIPBOARD_SHOW"],
    description="List all clipboard entries",
    validate=_validate,
    handler=_handler,
    examples=[],
)
